### PR TITLE
Change the locking scheme in DMFT to unlock in setoutputstreamstate

### PR DESCRIPTION
**Pull request by jiteshkris**
A change in device transform manager brought about a deadlock scenario in DMFT sample. The sample now behaves like we have in the rest of the pipeline i.e. we release the lock and wait for the operations on the input pins to complete. This hang is responsible for AMD hangs. Testing is in progress. This hang is probably also the hang that lenovo is hitting. This fix doesn't have the GPU leak while decoding. I am testing the changes with a logitech BRIO. This is also tested in the OS samples repo

### CODE REVIEW

File: avstream/sampledevicemft/multipinmft.cpp [hunk 1]
```diff
@@ -1509,7 +1512,10 @@ HRESULT CMultipinMft::ChangeMediaTypeEx(
         //
         //  The media type will be set on the input pin by the time we return from the wait
         //          
-        DMFTCHECKHR_GOTO(spinPin->WaitForSetInputPinMediaChange(), done);
+        m_critSec.Unlock();
```
Comment (Sudhaushu) [inline avstream/sampledevicemft/multipinmft.cpp:1515]: can this be done via {scope} instead of explicit unlock?

Comment (jiteshkris) [reply avstream/sampledevicemft/multipinmft.cpp:1515]: Not really. The operation is thankfully serialized by DMFT manager. We wait here till the input pin stream state completes, for which the DMFT will get called again

**Review by NaveenThum — APPROVED** [review]
:shipit:

**Comment by NaveenThum** [comment]
@chrilaMSFT is added to the review. #Closed

**Comment by NaveenThum** [comment]
@kasumman is added to the review. #Closed

**Comment by NaveenThum** [comment]
@SangChoe is added to the review. #Closed

**Comment by NaveenThum** [comment]
@strongstrongyiwen is added to the review. #Closed

**Comment by NaveenThum** [comment]
@absubba is added to the review. #Closed

**Comment by NaveenThum** [comment]
@LPBourret is added to the review. #Closed

**Comment by NaveenThum** [comment]
@gerritsMSFT is added to the review. #Closed

**Comment by NaveenThum** [comment]
@j-leungyy is added to the review. #Closed

File: avstream/sampledevicemft/common.h [hunk 2]
```diff
@@ -232,7 +232,7 @@ typedef enum _DMFT_conversion_type{
 
 typedef  std::vector< IMFMediaType *> IMFMediaTypeArray;
```
Comment (chrilaMSFT) [inline avstream/sampledevicemft/common.h:233]: >IMFMediaType [](http://example.com/codeflow?start=22&length=12)

Can you make all the vectors auto ref?

**Comment by chrilaMSFT** [comment]
>     ~CAsyncInPin()

virtual destructor?

---
Refers to: avstream/sampledevicemft/basepin.h:622 in e8ffee8. [](commit_id = e8ffee87d41a64e340e56d521a45bf03bd125502, deletion_comment = False)

**Comment by chrilaMSFT** [comment]
>     ~COutPin();

virtual destructor?

---
Refers to: avstream/sampledevicemft/basepin.h:544 in e8ffee8. [](commit_id = e8ffee87d41a64e340e56d521a45bf03bd125502, deletion_comment = False)

**Comment by chrilaMSFT** [comment]
>     )

Can you add the following pattern to all the methods, we are getting drivers that are letting exceptions escape their Functions

[https://www.osgwiki.com/wiki/Cpp_Core_Guidelines#E.MSFT.47:_Never_let_an_exception_escape_from_an_error_code_returning_method](https://www.osgwiki.com/wiki/Cpp_Core_Guidelines#E.MSFT.47:_Never_let_an_exception_escape_from_an_error_code_returning_method)


---
Refers to: avstream/sampledevicemft/multipinmft.cpp:1787 in e8ffee8. [](commit_id = e8ffee87d41a64e340e56d521a45bf03bd125502, deletion_comment = False)

File: avstream/sampledevicemft/multipinmfthelpers.cpp [hunk 3]
```diff
@@ -52,8 +51,6 @@ STDMETHODIMP_(VOID) CPinQueue::InsertInternal( _In_ IMFSample *pSample )
     if (FAILED(hr))
     {
         DMFTRACE(DMFT_GENERAL, TRACE_LEVEL_INFORMATION, "%!FUNC! exiting %x = %!HRESULT!", hr, hr);
-        // There is a bug in the pipeline that doesn't release the sample fed from processinput. We have to explicitly release the sample here
-        SAFE_RELEASE(pSample);
```
Comment (chrilaMSFT) [inline avstream/sampledevicemft/multipinmfthelpers.cpp:56]: > SAFE_RELEASE(pSample); [](http://example.com/codeflow?start=7&length=23)

Don't we still need to release the extra ref?

File: avstream/sampledevicemft/multipinmfthelpers.cpp [hunk 4]
```diff
@@ -1125,7 +1122,7 @@ VOID CDecoderTee::ShutdownTee()
 CGrayTee::CGrayTee(_In_ Ctee *tee) : CWrapTee(tee),m_transformfn(nullptr)
 {
 }
-STDMETHODIMP CGrayTee::Do(_In_ IMFSample *pSample, _Outptr_ IMFSample** ppOutSample, _Inout_ bool &newSample)
+STDMETHODIMP CGrayTee::Do(_In_ IMFSample *pSample, _Outptr_ IMFSample** ppOutSample)
 {
     HRESULT                 hr = S_OK;
```
Comment (chrilaMSFT) [inline avstream/sampledevicemft/multipinmfthelpers.cpp:1127]: NIT: Fix Spacing

**Comment by chrilaMSFT** [comment]
>     }

I think we are missing shutdown on the output pin as well.  It is present on the class it inherits from

---
Refers to: avstream/sampledevicemft/multipinmft.cpp:1746 in e8ffee8. [](commit_id = e8ffee87d41a64e340e56d521a45bf03bd125502, deletion_comment = False)

**Comment by chrilaMSFT** [comment]
> #if defined (MF_DEVICEMFT_ALLOW_MFT0_LOAD) && defined (MFT_UNIQUE_METHOD_NAMES)

Should we remove MFT0, this is not a supported scenario

---
Refers to: avstream/sampledevicemft/multipinmft.cpp:1742 in e8ffee8. [](commit_id = e8ffee87d41a64e340e56d521a45bf03bd125502, deletion_comment = False)

**Comment by NaveenThum** [comment]
> #if defined (MF_DEVICEMFT_ALLOW_MFT0_LOAD) && defined (MFT_UNIQUE_METHOD_NAMES)

There is a bug/task tracking removal of MFT0 from the sample

---
In reply to: [1349599849](https://github.com/microsoft/Windows-driver-samples/pull/810#issuecomment-1349599849) [](http://example.com/codeflow?ancestors=1349599849,1349599849)

---
Refers to: avstream/sampledevicemft/multipinmft.cpp:1742 in e8ffee8. [](commit_id = e8ffee87d41a64e340e56d521a45bf03bd125502, deletion_comment = False)

File: avstream/sampledevicemft/multipinmfthelpers.cpp (hunk 3 above)
Comment (jiteshkris) [reply avstream/sampledevicemft/multipinmfthelpers.cpp:56]: Nope. The sample is released in the input function i.e. ProcessInput. The other releases were just programming error which I could solve using a vecto<comptr> samplelist

File: avstream/sampledevicemft/common.h (hunk 2 above)
Comment (jiteshkris) [reply avstream/sampledevicemft/common.h:233]: Done

**Comment by sasaki0487** [comment]
Hi, I tried the code and modified basepin.h line 742 from 
```const GUID translatedGUID = MFVideoFormat_NV12;```
to
```const GUID translatedGUID = MFVideoFormat_YUY2;```
and the stream can't start
It worked for the older version of sampledevicemft, do i need to modify something else?

**Pull request merged**